// elk-message/dist/lib/enums/AlarmState.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AlarmState;
(function (AlarmState) {
    AlarmState[AlarmState["None"] = 48] = "None";
    AlarmState[AlarmState["EntranceDelayActive"] = 49] = "EntranceDelayActive";
    AlarmState[AlarmState["AlarmAbortDelayActive"] = 50] = "AlarmAbortDelayActive";
    AlarmState[AlarmState["FireAlarm"] = 51] = "FireAlarm";
    AlarmState[AlarmState["MedicalAlarm"] = 52] = "MedicalAlarm";
    AlarmState[AlarmState["PoliceAlarm"] = 53] = "PoliceAlarm";
    AlarmState[AlarmState["BurglarAlarm"] = 54] = "BurglarAlarm";
    AlarmState[AlarmState["Aux1Alarm"] = 55] = "Aux1Alarm";
    AlarmState[AlarmState["Aux2Alarm"] = 56] = "Aux2Alarm";
    AlarmState[AlarmState["Aux3Alarm"] = 57] = "Aux3Alarm";
    AlarmState[AlarmState["Aux4Alarm"] = 58] = "Aux4Alarm";
    AlarmState[AlarmState["CarbonMonoxideAlarm"] = 59] = "CarbonMonoxideAlarm";
    AlarmState[AlarmState["EmergencyAlarm"] = 60] = "EmergencyAlarm";
    AlarmState[AlarmState["FreezeAlarm"] = 61] = "FreezeAlarm";
    AlarmState[AlarmState["GasAlarm"] = 62] = "GasAlarm";
    AlarmState[AlarmState["HeatAlarm"] = 63] = "HeatAlarm";
    AlarmState[AlarmState["WaterAlarm"] = 64] = "WaterAlarm";
    AlarmState[AlarmState["FireSupervisory"] = 65] = "FireSupervisory";
    AlarmState[AlarmState["VerifyFire"] = 66] = "VerifyFire"; // 'B'
})(AlarmState || (AlarmState = {}));
(function (AlarmState) {
    /**
     * Determines if an AlarmState value indicates an active alarm
     * @param alarmState The AlarmState to check
     * @returns {boolean} True if the value is an active alarm;
     *   Otherwise, false.
     */
    function isActiveAlarm(alarmState) {
        // tslint:disable-next-line:no-unnecessary-qualifier
        return alarmState > AlarmState.AlarmAbortDelayActive;
    }
    AlarmState.isActiveAlarm = isActiveAlarm;
})(AlarmState || (AlarmState = {}));
exports.default = AlarmState;
