// elk-message/dist/lib/enums/ArmingLevel.js
"use strict";
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var ArmingLevel;
(function (ArmingLevel) {
    ArmingLevel[ArmingLevel["Disarm"] = 48] = "Disarm";
    ArmingLevel[ArmingLevel["ArmedAway"] = 49] = "ArmedAway";
    ArmingLevel[ArmingLevel["ArmedStay"] = 50] = "ArmedStay";
    ArmingLevel[ArmingLevel["ArmedStayInstant"] = 51] = "ArmedStayInstant";
    ArmingLevel[ArmingLevel["ArmedNight"] = 52] = "ArmedNight";
    ArmingLevel[ArmingLevel["ArmedNightInstant"] = 53] = "ArmedNightInstant";
    ArmingLevel[ArmingLevel["ArmedVacation"] = 54] = "ArmedVacation";
    ArmingLevel[ArmingLevel["ArmToNextAwayMode"] = 55] = "ArmToNextAwayMode";
    ArmingLevel[ArmingLevel["ArmToNextStayMode"] = 56] = "ArmToNextStayMode";
    ArmingLevel[ArmingLevel["ForceArmToAwayMode"] = 57] = "ForceArmToAwayMode";
    ArmingLevel[ArmingLevel["ForceArmToStayMode"] = 58] = "ForceArmToStayMode"; // ':', // Version 5.3.0 or later
})(ArmingLevel || (ArmingLevel = {}));
/* tslint:disable:no-unnecessary-qualifier */
(function (ArmingLevel) {
    ArmingLevel.AWAY_LEVELS = [
        ArmingLevel.ArmedAway,
        ArmingLevel.ArmedVacation,
        ArmingLevel.ArmToNextAwayMode,
        ArmingLevel.ForceArmToAwayMode
    ];
    ArmingLevel.NIGHT_LEVELS = [
        ArmingLevel.ArmedNight,
        ArmingLevel.ArmedNightInstant
    ];
    ArmingLevel.STAY_LEVELS = __spreadArrays(ArmingLevel.NIGHT_LEVELS, [
        ArmingLevel.ArmedStay,
        ArmingLevel.ArmedStayInstant,
        ArmingLevel.ArmToNextStayMode,
        ArmingLevel.ForceArmToStayMode
    ]);
    ArmingLevel.INSTANT_LEVELS = [
        ArmingLevel.ArmedStayInstant,
        ArmingLevel.ArmedNightInstant
    ];
    ArmingLevel.FORCE_LEVELS = [
        ArmingLevel.ForceArmToAwayMode,
        ArmingLevel.ForceArmToStayMode
    ];
    ArmingLevel.VACATION_LEVELS = [ArmingLevel.ArmedVacation];
    function isAway(armingLevel) {
        return ArmingLevel.AWAY_LEVELS.includes(armingLevel);
    }
    ArmingLevel.isAway = isAway;
    function isNight(armingLevel) {
        return ArmingLevel.NIGHT_LEVELS.includes(armingLevel);
    }
    ArmingLevel.isNight = isNight;
    function isStay(armingLevel) {
        return ArmingLevel.STAY_LEVELS.includes(armingLevel);
    }
    ArmingLevel.isStay = isStay;
    function isInstant(armingLevel) {
        return ArmingLevel.INSTANT_LEVELS.includes(armingLevel);
    }
    ArmingLevel.isInstant = isInstant;
    function isForce(armingLevel) {
        return ArmingLevel.FORCE_LEVELS.includes(armingLevel);
    }
    ArmingLevel.isForce = isForce;
    function isVacation(armingLevel) {
        return ArmingLevel.VACATION_LEVELS.includes(armingLevel);
    }
    ArmingLevel.isVacation = isVacation;
})(ArmingLevel || (ArmingLevel = {}));
exports.default = ArmingLevel;
