// elk-message/dist/lib/enums/AudioCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AudioCommand;
(function (AudioCommand) {
    AudioCommand[AudioCommand["PowerOff"] = 0] = "PowerOff";
    AudioCommand[AudioCommand["PowerOn"] = 1] = "PowerOn";
    AudioCommand[AudioCommand["PowerToggle"] = 2] = "PowerToggle";
    AudioCommand[AudioCommand["NextSource"] = 3] = "NextSource";
    AudioCommand[AudioCommand["Source"] = 4] = "Source";
    AudioCommand[AudioCommand["PreviousSelect"] = 5] = "PreviousSelect";
    AudioCommand[AudioCommand["NextSelect"] = 6] = "NextSelect";
    AudioCommand[AudioCommand["VolumeDown"] = 7] = "VolumeDown";
    AudioCommand[AudioCommand["VolumeUp"] = 8] = "VolumeUp";
    AudioCommand[AudioCommand["MuteAudio"] = 9] = "MuteAudio";
    AudioCommand[AudioCommand["VolumeSet"] = 10] = "VolumeSet";
    AudioCommand[AudioCommand["Play"] = 11] = "Play";
    AudioCommand[AudioCommand["Pause"] = 12] = "Pause";
    AudioCommand[AudioCommand["Stop"] = 13] = "Stop";
    AudioCommand[AudioCommand["SelectFavorite1"] = 14] = "SelectFavorite1";
    AudioCommand[AudioCommand["SelectFavorite2"] = 15] = "SelectFavorite2";
    AudioCommand[AudioCommand["Minus"] = 16] = "Minus";
    AudioCommand[AudioCommand["Plus"] = 17] = "Plus";
    AudioCommand[AudioCommand["AllZonesOff"] = 18] = "AllZonesOff";
    AudioCommand[AudioCommand["AllZonesOn"] = 19] = "AllZonesOn";
    AudioCommand[AudioCommand["AudioSystemManufacturer"] = 20] = "AudioSystemManufacturer";
})(AudioCommand || (AudioCommand = {}));
exports.default = AudioCommand;
