// elk-message/dist/lib/enums/EventType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var EventType;
(function (EventType) {
    EventType[EventType["NoEvent"] = 1000] = "NoEvent";
    EventType[EventType["FireALarm"] = 1001] = "FireALarm";
    EventType[EventType["FireSupervisoryAlarm"] = 1002] = "FireSupervisoryAlarm";
    EventType[EventType["BurglarAlarmAnyArea"] = 1003] = "BurglarAlarmAnyArea";
    EventType[EventType["MedicalAlarmAnyArea"] = 1004] = "MedicalAlarmAnyArea";
    EventType[EventType["PoliceAlarmAnyArea"] = 1005] = "PoliceAlarmAnyArea";
    EventType[EventType["Aux1_24HourAlarmAnyArea"] = 1006] = "Aux1_24HourAlarmAnyArea";
    EventType[EventType["Aux2_24HourAlarmAnyArea"] = 1007] = "Aux2_24HourAlarmAnyArea";
    EventType[EventType["CarbonMonoxideAlarmAnyArea"] = 1008] = "CarbonMonoxideAlarmAnyArea";
    EventType[EventType["EmergencyAlarmAnyArea"] = 1009] = "EmergencyAlarmAnyArea";
    EventType[EventType["FreezeAlarmAnyArea"] = 1010] = "FreezeAlarmAnyArea";
    EventType[EventType["GasAlarmAnyArea"] = 1011] = "GasAlarmAnyArea";
    EventType[EventType["HeatAlarmAnyArea"] = 1012] = "HeatAlarmAnyArea";
    EventType[EventType["WaterAlarmAnyArea"] = 1013] = "WaterAlarmAnyArea";
    EventType[EventType["AlarmAnyArea"] = 1014] = "AlarmAnyArea";
    EventType[EventType["BurglarAlarmArea1"] = 1015] = "BurglarAlarmArea1";
    EventType[EventType["BurglarAlarmArea2"] = 1016] = "BurglarAlarmArea2";
    EventType[EventType["BurglarAlarmArea3"] = 1017] = "BurglarAlarmArea3";
    EventType[EventType["BurglarAlarmArea4"] = 1018] = "BurglarAlarmArea4";
    EventType[EventType["BurglarAlarmArea5"] = 1019] = "BurglarAlarmArea5";
    EventType[EventType["BurglarAlarmArea6"] = 1020] = "BurglarAlarmArea6";
    EventType[EventType["BurglarAlarmArea7"] = 1021] = "BurglarAlarmArea7";
    EventType[EventType["BurglarAlarmArea8"] = 1022] = "BurglarAlarmArea8";
    EventType[EventType["MedicalAlarmArea1"] = 1023] = "MedicalAlarmArea1";
    EventType[EventType["MedicalAlarmArea2"] = 1024] = "MedicalAlarmArea2";
    EventType[EventType["MedicalAlarmArea3"] = 1025] = "MedicalAlarmArea3";
    EventType[EventType["MedicalAlarmArea4"] = 1026] = "MedicalAlarmArea4";
    EventType[EventType["MedicalAlarmArea5"] = 1027] = "MedicalAlarmArea5";
    EventType[EventType["MedicalAlarmArea6"] = 1028] = "MedicalAlarmArea6";
    EventType[EventType["MedicalAlarmArea7"] = 1029] = "MedicalAlarmArea7";
    EventType[EventType["MedicalAlarmArea8"] = 1030] = "MedicalAlarmArea8";
    EventType[EventType["PoliceAlarmArea1"] = 1031] = "PoliceAlarmArea1";
    EventType[EventType["PoliceAlarmArea2"] = 1032] = "PoliceAlarmArea2";
    EventType[EventType["PoliceAlarmArea3"] = 1033] = "PoliceAlarmArea3";
    EventType[EventType["PoliceAlarmArea4"] = 1034] = "PoliceAlarmArea4";
    EventType[EventType["PoliceAlarmArea5"] = 1035] = "PoliceAlarmArea5";
    EventType[EventType["PoliceAlarmArea6"] = 1036] = "PoliceAlarmArea6";
    EventType[EventType["PoliceAlarmArea7"] = 1037] = "PoliceAlarmArea7";
    EventType[EventType["PoliceAlarmArea8"] = 1038] = "PoliceAlarmArea8";
    EventType[EventType["Aux1_24HourAlarmArea1"] = 1039] = "Aux1_24HourAlarmArea1";
    EventType[EventType["Aux1_24HourAlarmArea2"] = 1040] = "Aux1_24HourAlarmArea2";
    EventType[EventType["Aux1_24HourAlarmArea3"] = 1041] = "Aux1_24HourAlarmArea3";
    EventType[EventType["Aux1_24HourAlarmArea4"] = 1042] = "Aux1_24HourAlarmArea4";
    EventType[EventType["Aux1_24HourAlarmArea5"] = 1043] = "Aux1_24HourAlarmArea5";
    EventType[EventType["Aux1_24HourAlarmArea6"] = 1044] = "Aux1_24HourAlarmArea6";
    EventType[EventType["Aux1_24HourAlarmArea7"] = 1045] = "Aux1_24HourAlarmArea7";
    EventType[EventType["Aux1_24HourAlarmArea8"] = 1046] = "Aux1_24HourAlarmArea8";
    EventType[EventType["Aux2_24HourAlarmArea1"] = 1047] = "Aux2_24HourAlarmArea1";
    EventType[EventType["Aux2_24HourAlarmArea2"] = 1048] = "Aux2_24HourAlarmArea2";
    EventType[EventType["Aux2_24HourAlarmArea3"] = 1049] = "Aux2_24HourAlarmArea3";
    EventType[EventType["Aux2_24HourAlarmArea4"] = 1050] = "Aux2_24HourAlarmArea4";
    EventType[EventType["Aux2_24HourAlarmArea5"] = 1051] = "Aux2_24HourAlarmArea5";
    EventType[EventType["Aux2_24HourAlarmArea6"] = 1052] = "Aux2_24HourAlarmArea6";
    EventType[EventType["Aux2_24HourAlarmArea7"] = 1053] = "Aux2_24HourAlarmArea7";
    EventType[EventType["Aux2_24HourAlarmArea8"] = 1054] = "Aux2_24HourAlarmArea8";
    EventType[EventType["CarbonMonoxideAlarmArea1"] = 1055] = "CarbonMonoxideAlarmArea1";
    EventType[EventType["CarbonMonoxideAlarmArea2"] = 1056] = "CarbonMonoxideAlarmArea2";
    EventType[EventType["CarbonMonoxideAlarmArea3"] = 1057] = "CarbonMonoxideAlarmArea3";
    EventType[EventType["CarbonMonoxideAlarmArea4"] = 1058] = "CarbonMonoxideAlarmArea4";
    EventType[EventType["CarbonMonoxideAlarmArea5"] = 1059] = "CarbonMonoxideAlarmArea5";
    EventType[EventType["CarbonMonoxideAlarmArea6"] = 1060] = "CarbonMonoxideAlarmArea6";
    EventType[EventType["CarbonMonoxideAlarmArea7"] = 1061] = "CarbonMonoxideAlarmArea7";
    EventType[EventType["CarbonMonoxideAlarmArea8"] = 1062] = "CarbonMonoxideAlarmArea8";
    EventType[EventType["EmergencyAlarmArea1"] = 1063] = "EmergencyAlarmArea1";
    EventType[EventType["EmergencyAlarmArea2"] = 1064] = "EmergencyAlarmArea2";
    EventType[EventType["EmergencyAlarmArea3"] = 1065] = "EmergencyAlarmArea3";
    EventType[EventType["EmergencyAlarmArea4"] = 1066] = "EmergencyAlarmArea4";
    EventType[EventType["EmergencyAlarmArea5"] = 1067] = "EmergencyAlarmArea5";
    EventType[EventType["EmergencyAlarmArea6"] = 1068] = "EmergencyAlarmArea6";
    EventType[EventType["EmergencyAlarmArea7"] = 1069] = "EmergencyAlarmArea7";
    EventType[EventType["EmergencyAlarmArea8"] = 1070] = "EmergencyAlarmArea8";
    EventType[EventType["FreezeAlarmArea1"] = 1071] = "FreezeAlarmArea1";
    EventType[EventType["FreezeAlarmArea2"] = 1072] = "FreezeAlarmArea2";
    EventType[EventType["FreezeAlarmArea3"] = 1073] = "FreezeAlarmArea3";
    EventType[EventType["FreezeAlarmArea4"] = 1074] = "FreezeAlarmArea4";
    EventType[EventType["FreezeAlarmArea5"] = 1075] = "FreezeAlarmArea5";
    EventType[EventType["FreezeAlarmArea6"] = 1076] = "FreezeAlarmArea6";
    EventType[EventType["FreezeAlarmArea7"] = 1077] = "FreezeAlarmArea7";
    EventType[EventType["FreezeAlarmArea8"] = 1078] = "FreezeAlarmArea8";
    EventType[EventType["GasAlarmArea1"] = 1079] = "GasAlarmArea1";
    EventType[EventType["GasAlarmArea2"] = 1080] = "GasAlarmArea2";
    EventType[EventType["GasAlarmArea3"] = 1081] = "GasAlarmArea3";
    EventType[EventType["GasAlarmArea4"] = 1082] = "GasAlarmArea4";
    EventType[EventType["GasAlarmArea5"] = 1083] = "GasAlarmArea5";
    EventType[EventType["GasAlarmArea6"] = 1084] = "GasAlarmArea6";
    EventType[EventType["GasAlarmArea7"] = 1085] = "GasAlarmArea7";
    EventType[EventType["GasAlarmArea8"] = 1086] = "GasAlarmArea8";
    EventType[EventType["HeatAlarmArea1"] = 1087] = "HeatAlarmArea1";
    EventType[EventType["HeatAlarmArea2"] = 1088] = "HeatAlarmArea2";
    EventType[EventType["HeatAlarmArea3"] = 1089] = "HeatAlarmArea3";
    EventType[EventType["HeatAlarmArea4"] = 1090] = "HeatAlarmArea4";
    EventType[EventType["HeatAlarmArea5"] = 1091] = "HeatAlarmArea5";
    EventType[EventType["HeatAlarmArea6"] = 1092] = "HeatAlarmArea6";
    EventType[EventType["HeatAlarmArea7"] = 1093] = "HeatAlarmArea7";
    EventType[EventType["HeatAlarmArea8"] = 1094] = "HeatAlarmArea8";
    EventType[EventType["WaterAlarmArea1"] = 1095] = "WaterAlarmArea1";
    EventType[EventType["WaterAlarmArea2"] = 1096] = "WaterAlarmArea2";
    EventType[EventType["WaterAlarmArea3"] = 1097] = "WaterAlarmArea3";
    EventType[EventType["WaterAlarmArea4"] = 1098] = "WaterAlarmArea4";
    EventType[EventType["WaterAlarmArea5"] = 1099] = "WaterAlarmArea5";
    EventType[EventType["WaterAlarmArea6"] = 1100] = "WaterAlarmArea6";
    EventType[EventType["WaterAlarmArea7"] = 1101] = "WaterAlarmArea7";
    EventType[EventType["WaterAlarmArea8"] = 1102] = "WaterAlarmArea8";
    EventType[EventType["AlarmArea1"] = 1103] = "AlarmArea1";
    EventType[EventType["AlarmArea2"] = 1104] = "AlarmArea2";
    EventType[EventType["AlarmArea3"] = 1105] = "AlarmArea3";
    EventType[EventType["AlarmArea4"] = 1106] = "AlarmArea4";
    EventType[EventType["AlarmArea5"] = 1107] = "AlarmArea5";
    EventType[EventType["AlarmArea6"] = 1108] = "AlarmArea6";
    EventType[EventType["AlarmArea7"] = 1109] = "AlarmArea7";
    EventType[EventType["AlarmArea8"] = 1110] = "AlarmArea8";
    EventType[EventType["CodeLockout"] = 1111] = "CodeLockout";
    EventType[EventType["CodeLockoutKeypad1"] = 1112] = "CodeLockoutKeypad1";
    EventType[EventType["CodeLockoutKeypad2"] = 1113] = "CodeLockoutKeypad2";
    EventType[EventType["CodeLockoutKeypad3"] = 1114] = "CodeLockoutKeypad3";
    EventType[EventType["CodeLockoutKeypad4"] = 1115] = "CodeLockoutKeypad4";
    EventType[EventType["CodeLockoutKeypad5"] = 1116] = "CodeLockoutKeypad5";
    EventType[EventType["CodeLockoutKeypad6"] = 1117] = "CodeLockoutKeypad6";
    EventType[EventType["CodeLockoutKeypad7"] = 1118] = "CodeLockoutKeypad7";
    EventType[EventType["CodeLockoutKeypad8"] = 1119] = "CodeLockoutKeypad8";
    EventType[EventType["CodeLockoutKeypad9"] = 1120] = "CodeLockoutKeypad9";
    EventType[EventType["CodeLockoutKeypad10"] = 1121] = "CodeLockoutKeypad10";
    EventType[EventType["CodeLockoutKeypad11"] = 1122] = "CodeLockoutKeypad11";
    EventType[EventType["CodeLockoutKeypad12"] = 1123] = "CodeLockoutKeypad12";
    EventType[EventType["CodeLockoutKeypad13"] = 1124] = "CodeLockoutKeypad13";
    EventType[EventType["CodeLockoutKeypad14"] = 1125] = "CodeLockoutKeypad14";
    EventType[EventType["CodeLockoutKeypad15"] = 1126] = "CodeLockoutKeypad15";
    EventType[EventType["CodeLockoutKeypad16"] = 1127] = "CodeLockoutKeypad16";
    EventType[EventType["FireTrouble"] = 1128] = "FireTrouble";
    EventType[EventType["BurglarTrouble"] = 1129] = "BurglarTrouble";
    EventType[EventType["FailToCommunicateTrouble"] = 1130] = "FailToCommunicateTrouble";
    EventType[EventType["RfSensorLowBatteryTrouble"] = 1131] = "RfSensorLowBatteryTrouble";
    EventType[EventType["LostAncModuleTrouble"] = 1132] = "LostAncModuleTrouble";
    EventType[EventType["LostKeypadTrouble"] = 1133] = "LostKeypadTrouble";
    EventType[EventType["LostInputExpanderTrouble"] = 1134] = "LostInputExpanderTrouble";
    EventType[EventType["LostOutputExpanderTrouble"] = 1135] = "LostOutputExpanderTrouble";
})(EventType || (EventType = {}));
exports.default = EventType;
