// elk-message/dist/lib/enums/KeypadKey.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var KeypadKey;
(function (KeypadKey) {
    KeypadKey[KeypadKey["None"] = 0] = "None";
    KeypadKey[KeypadKey["Star"] = 11] = "Star";
    KeypadKey[KeypadKey["Pound"] = 12] = "Pound";
    KeypadKey[KeypadKey["F1"] = 13] = "F1";
    KeypadKey[KeypadKey["F2"] = 14] = "F2";
    KeypadKey[KeypadKey["F3"] = 15] = "F3";
    KeypadKey[KeypadKey["F4"] = 16] = "F4";
    KeypadKey[KeypadKey["Stay"] = 17] = "Stay";
    KeypadKey[KeypadKey["Exit"] = 18] = "Exit";
    KeypadKey[KeypadKey["Chime"] = 19] = "Chime";
    KeypadKey[KeypadKey["Bypass"] = 20] = "Bypass";
    KeypadKey[KeypadKey["Elk"] = 21] = "Elk";
    KeypadKey[KeypadKey["Down"] = 22] = "Down";
    KeypadKey[KeypadKey["Up"] = 23] = "Up";
    KeypadKey[KeypadKey["Right"] = 24] = "Right";
    KeypadKey[KeypadKey["Left"] = 25] = "Left";
    KeypadKey[KeypadKey["F6"] = 26] = "F6";
    KeypadKey[KeypadKey["F7"] = 27] = "F7";
    KeypadKey[KeypadKey["Data"] = 28] = "Data"; // Data was entered, this acts as a carriage return
})(KeypadKey || (KeypadKey = {}));
exports.default = KeypadKey;
