// elk-message/dist/lib/enums/PlcFunctionCode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var PlcFunctionCode;
(function (PlcFunctionCode) {
    PlcFunctionCode[PlcFunctionCode["AllUnitsOff"] = 1] = "AllUnitsOff";
    PlcFunctionCode[PlcFunctionCode["AllLightsOn"] = 2] = "AllLightsOn";
    PlcFunctionCode[PlcFunctionCode["UnitOn"] = 3] = "UnitOn";
    PlcFunctionCode[PlcFunctionCode["UnitOff"] = 4] = "UnitOff";
    PlcFunctionCode[PlcFunctionCode["Dim"] = 5] = "Dim";
    PlcFunctionCode[PlcFunctionCode["Bright"] = 6] = "Bright";
    PlcFunctionCode[PlcFunctionCode["AllLightsOff"] = 7] = "AllLightsOff";
    PlcFunctionCode[PlcFunctionCode["ExtendedCode"] = 8] = "ExtendedCode";
    PlcFunctionCode[PlcFunctionCode["PresetDim"] = 9] = "PresetDim";
    PlcFunctionCode[PlcFunctionCode["ExtendedData"] = 10] = "ExtendedData";
    PlcFunctionCode[PlcFunctionCode["StatusRequest"] = 11] = "StatusRequest";
    PlcFunctionCode[PlcFunctionCode["HailRequest"] = 12] = "HailRequest";
    PlcFunctionCode[PlcFunctionCode["HailAcknowledgement"] = 13] = "HailAcknowledgement";
    PlcFunctionCode[PlcFunctionCode["StatusOn"] = 14] = "StatusOn";
    PlcFunctionCode[PlcFunctionCode["StatusOff"] = 15] = "StatusOff";
})(PlcFunctionCode || (PlcFunctionCode = {}));
exports.default = PlcFunctionCode;
