// elk-message/dist/lib/enums/SystemTroubleType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var SystemTroubleType;
(function (SystemTroubleType) {
    SystemTroubleType[SystemTroubleType["None"] = 0] = "None";
    SystemTroubleType[SystemTroubleType["ACFail"] = 1] = "ACFail";
    SystemTroubleType[SystemTroubleType["BoxTamper"] = 2] = "BoxTamper";
    SystemTroubleType[SystemTroubleType["FailToCommunicate"] = 4] = "FailToCommunicate";
    SystemTroubleType[SystemTroubleType["EEPromMemoryError"] = 8] = "EEPromMemoryError";
    SystemTroubleType[SystemTroubleType["LowBatteryControl"] = 16] = "LowBatteryControl";
    SystemTroubleType[SystemTroubleType["TransmitterLowBattery"] = 32] = "TransmitterLowBattery";
    SystemTroubleType[SystemTroubleType["OverCurrent"] = 64] = "OverCurrent";
    SystemTroubleType[SystemTroubleType["TelephoneFault"] = 128] = "TelephoneFault";
    SystemTroubleType[SystemTroubleType["Output2"] = 256] = "Output2";
    SystemTroubleType[SystemTroubleType["MissingKeypad"] = 512] = "MissingKeypad";
    SystemTroubleType[SystemTroubleType["ZoneExpander"] = 1024] = "ZoneExpander";
    SystemTroubleType[SystemTroubleType["OutputExpander"] = 2048] = "OutputExpander";
    SystemTroubleType[SystemTroubleType["ELKRPRemoteAccess"] = 4096] = "ELKRPRemoteAccess";
    SystemTroubleType[SystemTroubleType["CommonAreaNotArmed"] = 8192] = "CommonAreaNotArmed";
    SystemTroubleType[SystemTroubleType["FlashMemoryError"] = 16384] = "FlashMemoryError";
    SystemTroubleType[SystemTroubleType["SecurityAlert"] = 32768] = "SecurityAlert";
    SystemTroubleType[SystemTroubleType["SerialPortExpander"] = 65536] = "SerialPortExpander";
    SystemTroubleType[SystemTroubleType["LostTransmitter"] = 131072] = "LostTransmitter";
    SystemTroubleType[SystemTroubleType["GESmokeCleanMe"] = 262144] = "GESmokeCleanMe";
    SystemTroubleType[SystemTroubleType["EthernetTrouble"] = 524288] = "EthernetTrouble";
    SystemTroubleType[SystemTroubleType["DisplayMessageInKeypadLine1"] = 1048576] = "DisplayMessageInKeypadLine1";
    SystemTroubleType[SystemTroubleType["DisplayMessageInKeypadLine2"] = 2097152] = "DisplayMessageInKeypadLine2";
    SystemTroubleType[SystemTroubleType["FireTrouble"] = 4194304] = "FireTrouble";
})(SystemTroubleType || (SystemTroubleType = {}));
exports.default = SystemTroubleType;
