// elk-message/dist/lib/enums/TextDescriptionType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var TextDescriptionType;
(function (TextDescriptionType) {
    TextDescriptionType[TextDescriptionType["ZoneName"] = 0] = "ZoneName";
    TextDescriptionType[TextDescriptionType["AreaName"] = 1] = "AreaName";
    TextDescriptionType[TextDescriptionType["UserName"] = 2] = "UserName";
    TextDescriptionType[TextDescriptionType["KeypadName"] = 3] = "KeypadName";
    TextDescriptionType[TextDescriptionType["OutputName"] = 4] = "OutputName";
    TextDescriptionType[TextDescriptionType["TaskName"] = 5] = "TaskName";
    TextDescriptionType[TextDescriptionType["TelephoneName"] = 6] = "TelephoneName";
    TextDescriptionType[TextDescriptionType["LightName"] = 7] = "LightName";
    TextDescriptionType[TextDescriptionType["AlarmDurationName"] = 8] = "AlarmDurationName";
    TextDescriptionType[TextDescriptionType["CustomSettings"] = 9] = "CustomSettings";
    TextDescriptionType[TextDescriptionType["CounterName"] = 10] = "CounterName";
    TextDescriptionType[TextDescriptionType["ThermostatName"] = 11] = "ThermostatName";
    TextDescriptionType[TextDescriptionType["FunctionKey1Name"] = 12] = "FunctionKey1Name";
    TextDescriptionType[TextDescriptionType["FunctionKey2Name"] = 13] = "FunctionKey2Name";
    TextDescriptionType[TextDescriptionType["FunctionKey3Name"] = 14] = "FunctionKey3Name";
    TextDescriptionType[TextDescriptionType["FunctionKey4Name"] = 15] = "FunctionKey4Name";
    TextDescriptionType[TextDescriptionType["FunctionKey5Name"] = 16] = "FunctionKey5Name";
    TextDescriptionType[TextDescriptionType["FunctionKey6Name"] = 17] = "FunctionKey6Name";
    TextDescriptionType[TextDescriptionType["AudioZoneName"] = 18] = "AudioZoneName";
    TextDescriptionType[TextDescriptionType["AudioSourceName"] = 19] = "AudioSourceName";
})(TextDescriptionType || (TextDescriptionType = {}));
exports.default = TextDescriptionType;
