// elk-message/dist/lib/enums/ZoneDefinition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ZoneDefinition;
(function (ZoneDefinition) {
    ZoneDefinition[ZoneDefinition["Disabled"] = 48] = "Disabled";
    ZoneDefinition[ZoneDefinition["BurglarEntryExit1"] = 49] = "BurglarEntryExit1";
    ZoneDefinition[ZoneDefinition["BurglarEntryExit2"] = 50] = "BurglarEntryExit2";
    ZoneDefinition[ZoneDefinition["BurglarPerimeterInstant"] = 51] = "BurglarPerimeterInstant";
    ZoneDefinition[ZoneDefinition["BurglarInterior"] = 52] = "BurglarInterior";
    ZoneDefinition[ZoneDefinition["BurglarInteriorFollower"] = 53] = "BurglarInteriorFollower";
    ZoneDefinition[ZoneDefinition["BurglarInteriorNight"] = 54] = "BurglarInteriorNight";
    ZoneDefinition[ZoneDefinition["BurglarInteriorNightDelay"] = 55] = "BurglarInteriorNightDelay";
    ZoneDefinition[ZoneDefinition["Burglar24Hour"] = 56] = "Burglar24Hour";
    ZoneDefinition[ZoneDefinition["BurglarBoxTamper"] = 57] = "BurglarBoxTamper";
    ZoneDefinition[ZoneDefinition["FireAlarm"] = 58] = "FireAlarm";
    ZoneDefinition[ZoneDefinition["FireVerified"] = 59] = "FireVerified";
    ZoneDefinition[ZoneDefinition["FireSupervisory"] = 60] = "FireSupervisory";
    ZoneDefinition[ZoneDefinition["AuxAlarm1"] = 61] = "AuxAlarm1";
    ZoneDefinition[ZoneDefinition["AuxAlarm2"] = 62] = "AuxAlarm2";
    ZoneDefinition[ZoneDefinition["Keyfob"] = 63] = "Keyfob";
    ZoneDefinition[ZoneDefinition["NonAlarm"] = 64] = "NonAlarm";
    ZoneDefinition[ZoneDefinition["CarbonMonoxide"] = 65] = "CarbonMonoxide";
    ZoneDefinition[ZoneDefinition["EmergencyAlarm"] = 66] = "EmergencyAlarm";
    ZoneDefinition[ZoneDefinition["FreezeAlarm"] = 67] = "FreezeAlarm";
    ZoneDefinition[ZoneDefinition["GasAlarm"] = 68] = "GasAlarm";
    ZoneDefinition[ZoneDefinition["HeatAlarm"] = 69] = "HeatAlarm";
    ZoneDefinition[ZoneDefinition["MedicalAlarm"] = 70] = "MedicalAlarm";
    ZoneDefinition[ZoneDefinition["PoliceAlarm"] = 71] = "PoliceAlarm";
    ZoneDefinition[ZoneDefinition["PoliceNoIndication"] = 72] = "PoliceNoIndication";
    ZoneDefinition[ZoneDefinition["WaterAlarm"] = 73] = "WaterAlarm";
    ZoneDefinition[ZoneDefinition["KeyMomentaryArmDisarm"] = 74] = "KeyMomentaryArmDisarm";
    ZoneDefinition[ZoneDefinition["KeyMomentaryArmAway"] = 75] = "KeyMomentaryArmAway";
    ZoneDefinition[ZoneDefinition["KeyMomentaryArmStay"] = 76] = "KeyMomentaryArmStay";
    ZoneDefinition[ZoneDefinition["KeyMomentaryDisarm"] = 77] = "KeyMomentaryDisarm";
    ZoneDefinition[ZoneDefinition["KeyOnOff"] = 78] = "KeyOnOff";
    ZoneDefinition[ZoneDefinition["MuteAudibles"] = 79] = "MuteAudibles";
    ZoneDefinition[ZoneDefinition["PowerSupervisory"] = 80] = "PowerSupervisory";
    ZoneDefinition[ZoneDefinition["Temperature"] = 81] = "Temperature";
    ZoneDefinition[ZoneDefinition["AnalogZone"] = 82] = "AnalogZone";
    ZoneDefinition[ZoneDefinition["PhoneKey"] = 83] = "PhoneKey";
    ZoneDefinition[ZoneDefinition["IntercomKey"] = 84] = "IntercomKey"; // 'T'
})(ZoneDefinition || (ZoneDefinition = {}));
exports.default = ZoneDefinition;
