// elk-client/dist/lib/connection/ElkSocketConnectionOptions.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * The default host name to use when one is not specified.
 *
 * 192.168.0.251 is the default IP address that the Elk M1
 * assigns to itself when it is not given one by a DHCP server
 * or when it is expicitly reset.
 */
exports.DEFAULT_HOST = '192.168.0.251';
/**
 * The default port to connect to when using an insecure connection.
 */
exports.DEFAULT_INSECURE_PORT = 2101;
/**
 * The default port to connect to when using a secure connection.
 */
exports.DEFAULT_SECURE_PORT = 2601;
/**
 * The default connection options to use if `secure` === false
 */
exports.DEFAULT_OPTIONS = {
    host: exports.DEFAULT_HOST,
    port: exports.DEFAULT_INSECURE_PORT,
    secure: false,
    idleTimeout: 60 * 1000,
    connectTimeout: 30 * 1000,
};
/**
 * The default connection options to use if `secure` === true
 */
exports.SECURE_DEFAULT_OPTIONS = __assign(__assign({}, exports.DEFAULT_OPTIONS), { secure: true, port: exports.DEFAULT_SECURE_PORT });
