// elk-client/dist/lib/errors/AuthenticationFailedError.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var AuthenticationFailedReason;
(function (AuthenticationFailedReason) {
    AuthenticationFailedReason[AuthenticationFailedReason["Unknown"] = 0] = "Unknown";
    AuthenticationFailedReason[AuthenticationFailedReason["InvalidCredentials"] = 1] = "InvalidCredentials";
    AuthenticationFailedReason[AuthenticationFailedReason["MissingUsername"] = 2] = "MissingUsername";
    AuthenticationFailedReason[AuthenticationFailedReason["MissingPassword"] = 3] = "MissingPassword";
})(AuthenticationFailedReason = exports.AuthenticationFailedReason || (exports.AuthenticationFailedReason = {}));
/**
 * Inidicates that an authentication error occured. Either
 * authentication failed, or there was no authentication
 * information provided.
 */
var AuthenticationFailedError = /** @class */ (function (_super) {
    __extends(AuthenticationFailedError, _super);
    function AuthenticationFailedError(reason, message) {
        if (reason === void 0) { reason = AuthenticationFailedReason.Unknown; }
        var _this = _super.call(this, message) || this;
        _this.reason = reason;
        Object.setPrototypeOf(_this, AuthenticationFailedError.prototype);
        return _this;
    }
    return AuthenticationFailedError;
}(Error));
exports.default = AuthenticationFailedError;
